qiskit/transpiler/passes/layout/disjoint_utils.py

```diff
@@ -0,0 +1,162 @@
+        out_dag = dag_components[dags.pop()]
+        for dag_index in dags:
+            dag = dag_components[dag_index]
+            out_dag.add_qubits(dag.qubits)
+            out_dag.add_clbits(dag.clbits)
+            out_dag.add_qreg(dag.qregs)
+            out_dag.add_cregs(dag.cregs)
+            out_dag.compose(dag, qubits=dag.qubits, clbits=dag.clbits)
+        out_component_pairs.append((out_dag, cmap_components[cmap_index]))
+    res = [run_func(out_dag, cmap) for out_dag, cmap in out_component_pairs]
+    return res
+
+
+def map_components(
+    dag_components: List[DAGCircuit], cmap_components: List[CouplingMap]
+) -> Dict[int, int]:
+    """Map a list of circuit components to coupling map components."""
+    free_qubits = {index: len(cmap.graph) for index, cmap in enumerate(cmap_components)}
+    out_mapping = defaultdict(list)
+
+    for dag_index, dag in sorted(
+        enumerate(dag_components), key=lambda x: x[1].num_qubits(), reverse=True
+    ):
+        for cmap_index in sorted(
+            range(len(cmap_components)), key=lambda index: free_qubits[index], reverse=True
+        ):
+            # TODO: Improve heuristic to involve connectivity and estimate
+            # swap cost
+            if dag.num_qubits() <= free_qubits[cmap_index]:
+                out_mapping[cmap_index].append(dag_index)
+                free_qubits[cmap_index] -= dag.num_qubits()
+                break
+        else:
+            raise TranspilerError(
+                "A connected component of the DAGCircuit is too large for any of the connected "
+                "components in the coupling map."
+            )
+    return out_mapping
+
+
+def split_barriers(dag: DAGCircuit):
+    """Mutate an input dag to split barriers into single qubit barriers."""
+    for node in dag.op_nodes(Barrier):
+        num_qubits = len(node.qargs)
+        if num_qubits == 1:
+            continue
+        barrier_uuid = uuid.uuid4()
+        split_dag = DAGCircuit()
+        split_dag.add_qubits([Qubit() for _ in range(num_qubits)])
+        for i in range(num_qubits):
+            split_dag.apply_operation_back(
+                Barrier(1, label=barrier_uuid), qargs=[split_dag.qubits[i]]
+            )
+        dag.substitute_node_with_dag(node, split_dag)
+
+
+def combine_barriers(dag: DAGCircuit, retain_uuid: bool = True):
+    """Mutate input dag to combine barriers with UUID labels into a single barrier."""
+    qubit_indices = {bit: index for index, bit in enumerate(dag.qubits)}
+    uuid_map = {}
+    for node in dag.op_nodes(Barrier):
+        if isinstance(node.op.label, uuid.UUID):
+            barrier_uuid = node.op.label
+            if barrier_uuid in uuid_map:
+                other_node = uuid_map[node.op.label]
+                num_qubits = len(other_node.qargs) + len(node.qargs)
+                new_op = Barrier(num_qubits, label=barrier_uuid)
+                new_node = dag.replace_block_with_op([node, other_node], new_op, qubit_indices)
```

Comment:
I am potentially worried about this potentially introducing a cycle when combining a barrier across qubits. I think it's possible this could cause that which is why I left the cycle check enabled for this. But we should try to confirm this before merging.